src/amoc_scenario.erl

```diff
@@ -12,12 +27,128 @@
 -callback start(user_id()) -> any().
--callback continue() -> continue | {stop, Reason :: term()}.
--callback terminate(Reason :: term()) -> any().
--callback next_user_batch(BatchIndex :: non_neg_integer(),
-                          PrevUserCount :: non_neg_integer()) ->
-    amoc_controller:user_batch_strategy().
 
 %% either start/1 or start/2 must be exported from the behaviour module
 -optional_callbacks([start/1, start/2]).
--optional_callbacks([continue/0, terminate/1, next_user_batch/2]).
+
+%%-------------------------------------------------------------------------
+%% API
+%%-------------------------------------------------------------------------
+-spec start_link() -> {ok, pid()} | ignore | {error, term()}.
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+-spec install_scenario(module(), binary()) ->
+    ok | {error, [Errors :: string()], [Warnings :: string()]}.
+install_scenario(Module, ModuleSource) ->
+    gen_server:call(?MODULE, {add_scenario, Module, ModuleSource}).
+
+-spec does_scenario_exist(module()) -> boolean().
+does_scenario_exist(Scenario) ->
+    AllScenarios = list_scenario_modules(),
+    lists:member(Scenario, AllScenarios).
+
+-spec list_scenario_modules() -> [module()].
+list_scenario_modules() ->
+    [element(1, T) || T <- ets:tab2list(amoc_scenarios)].
+
+%%-------------------------------------------------------------------------
+%% gen_server callbacks
+%%-------------------------------------------------------------------------
+-spec init([]) -> {ok, state()}.
+init([]) ->
+    code:add_pathsz(?EBIN_DIR),
+    start_scenarios_ets(),
+    find_scenario_modules(),
+    {ok, ok}.
+
+-spec handle_call(any(), any(), state()) -> {reply, any(), state()}.
+handle_call({add_scenario, Module, ModuleSource}, _, State) ->
+    Reply = add_scenario(Module, ModuleSource),
+    {reply, Reply, State};
+handle_call(_, _, State) ->
+    {reply, {error, not_implemented}, State}.
+
+-spec handle_cast(any(), state()) -> {noreply, state()}.
+handle_cast(_, State) ->
+    {noreply, State}.
+
+%%-------------------------------------------------------------------------
+%%  local functions
+%%-------------------------------------------------------------------------
+-spec start_scenarios_ets() -> amoc_scenarios.
+start_scenarios_ets() ->
+    ets:new(amoc_scenarios, [named_table,
+                             protected,
+                             {read_concurrency, true}]).
+
+-spec find_scenario_modules() -> [module()].
+find_scenario_modules() ->
+    AllBeamFiles = [File || Path <- code:get_path(), File <- filelib:wildcard("*.beam", Path)],
+    AllModules = [list_to_atom(filename:rootname(BeamFile)) || BeamFile <- AllBeamFiles],
+    ok = code:ensure_modules_loaded(AllModules),
+    AllScenarios = [Module || Module <- erlang:loaded(), is_scenario(Module)],
+    [ets:insert(amoc_scenarios, {Scenario, scenario, preloaded}) || Scenario <- AllScenarios].
```

Comment:
Why do we need to have all possible scenarios preloaded?

Reply:
it's required for dynamic addition of the new amoc node during the scenario execution